src/cdomains/affineEquality/sparseImplementation/sparseVector.ml

```diff
@@ -0,0 +1,393 @@
+open AbstractVector
+open RatOps
+open ConvenienceOps
+
+open BatList
+open Batteries
+
+module List = BatList
+module Array = BatArray
+
+module SparseVector: AbstractVector =
+  functor (A: RatOps) ->
+  struct
+    include ConvenienceOps (A)
+
+    type t = {
+      entries: (int * A.t) list ;
+      len: int
+    }[@@deriving eq, ord, hash]
+
+    let copy v = v 
+
+    let of_list l = 
+      let entries' = List.filteri_map (fun i x -> if x <> A.zero then Some (i, x) else None) l
+      in {entries = entries'; len = List.length l}
+
+    let of_array a =
+      let len' = Array.length a in
+      let entries' = List.rev @@ Array.fold_lefti (fun acc i x -> if x <> A.zero then (i, x) :: acc else acc ) [] a in
+      {entries = entries'; len = len'}
+
+    let of_sparse_list count ls =
+      {entries = ls; len = count}
+
+    (** [to_list v] Returns a non-sparse list representing the vector v *)
+    let to_list v =
+      let rec extend_zero_aux i v' acc =
+        match v' with
+        | (index, value) :: xs -> if index = i then extend_zero_aux (i + 1) xs (value :: acc) else extend_zero_aux (i + 1) v' (A.zero :: acc)
+        | [] -> if i < v.len then extend_zero_aux (i + 1) v' (A.zero :: acc) else List.rev acc
+      in
```

Comment:
```suggestion
      let rec extend_zero_aux i acc = function
          | (index, value) :: xs when index = i -> extend_zero_aux (i + 1) (value :: acc) xs
          | ((_:: _) as v) -> extend_zero_aux (i + 1) (A.zero :: acc) v
          | [] -> if i < v.len then extend_zero_aux (i + 1) (A.zero :: acc) [] else List.rev acc
      in
```

Not sure if this makes it easier to understand, but maybe yes? Please test before committing, not sure if it's 100% correct